### PR TITLE
fix(config): Make print-config not panic without a config

### LINKED ISSUE #4958: `starship print-config` panics without a configuration file
#### Current Behavior

In v1.13.1, `starship print-config` generates a traceback when there is no configuration file.

```shell
starship-1.13.1 (fdd3d15) is 📦 v1.13.1 via 🦀 v1.65.0 took 2m15s
❯ cargo run print-config
    Finished dev [unoptimized + debuginfo] target(s) in 0.16s
     Running `target/debug/starship print-config`
thread 'main' panicked at 'Failed to load starship config', src/configure.rs:225:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

With `RUST_BACKTRACE=1`:

```shell
starship-1.13.1 (fdd3d15) is 📦 v1.13.1 via 🦀 v1.65.0
❯ RUST_BACKTRACE=1 cargo run print-config
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
     Running `target/debug/starship print-config`
thread 'main' panicked at 'Failed to load starship config', src/configure.rs:225:10
stack backtrace:
   0: rust_begin_unwind
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
   2: core::panicking::panic_display
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:72:5
   3: core::panicking::panic_str
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:56:5
   4: core::option::expect_failed
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/option.rs:1880:5
   5: core::option::Option<T>::expect
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/option.rs:738:21
   6: starship::configure::get_configuration
             at ./src/configure.rs:223:5
   7: starship::configure::print_configuration
             at ./src/configure.rs:82:27
   8: starship::main
             at ./src/main.rs:216:52
   9: core::ops::function::FnOnce::call_once
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

#### Expected Behavior

No traceback was expected.  Previously in v1.12.0, no traceback occurs.

#### Additional context/Screenshots

In v1.12.0, `StarshipConfig::initialize()` was:

```rust
impl StarshipConfig {
    /// Initialize the Config struct
    pub fn initialize() -> Self {
        if let Some(file_data) = Self::config_from_file() {
            Self {
                config: Some(file_data),
            }
        } else {
            Self {
                config: Some(Value::Table(toml::value::Table::new())),
            }
        }
    }
    // ...
}
```

Without a configuration file, `StarshipConfig::config` was set to `Some(default_value)`.

In v1.13.1, `StarshipConfig::initialize()` is:

```rust
impl StarshipConfig {
    /// Initialize the Config struct
    pub fn initialize() -> Self {
        Self::config_from_file()
            .map(|config| Self {
                config: Some(config),
            })
            .unwrap_or_default()
    }
    // ...
}
```

Now, without a configuration file, `config_from_file()` will be `None`, the `.map()` leaves it `None`, and `.unwrap_or_default()` then returns the derived `Default` value for `StarshipConfig`:

```rust
/// Root config of starship.
#[derive(Default)]
pub struct StarshipConfig {
    pub config: Option<toml::Table>,
}
```

The default for an `Option` is `None`, so the default-generated `StarshipConfig::config` is `None`, causing the traceback.

#### Environment
- Starship version: 1.13.1
- bash version: GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
- Operating system: Ubuntu 22.04
- Terminal emulator: Plasma Konsole Version 21.12.3
- Git Commit Hash:
- Branch/Tag: v1.13.1
- Rust Version: rustc 1.65.0 (897e37553 2022-11-02)
- Rust channel: stable-x86_64-unknown-linux-gnu debug
- Build Time: 2023-03-03 18:03:02 -05:00

#### Starship Configuration

No starship configuration file is used.

### PR DESCRIPTION
**Pull request by domsleee**
#### Description
The problems are:
* `starship print-config` panics if no config exists
* `starship print-config` panics if an invalid `STARSHIP_CONFIG` env var is set

#### Motivation and Context

Closes #4958
Closes #5000

#### How Has This Been Tested?
I tested "no home, no env uses default" and "home, nonexisting env uses default" by running the binary.
It is particularly surprising that if you specify a nonexisting file in `STARSHIP_CONFIG`, the default config is used in the prompt, and not `~/.config/starship.toml`.
With this change, `print-config` is consistent with prompt behaviour I think.

- [x] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [x] I have tested using **Windows**

#### Checklist:

- [x] I have updated the documentation accordingly.
Current docs are correct "print-config: Prints the computed starship configuration"
- [x] I have updated the tests accordingly.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -125,6 +125,7 @@ winres = "0.1.12"
 
 [dev-dependencies]
 mockall = "0.11"
+temp-env = "0.3.2"
```
[the remark this answers is not in the document]
Comment (domsleee) [reply Cargo.toml:128]: Ah alright
I'm thinking a few things then:
* A new method `Context#get_config_path`
* Would need `Context#new_with_env` for testing since the init of `Context` uses STARSHIP_CONFIG / HOME on `new`
* There is `config.rs#get_config_path` and `configure.rs#get_config_path`. They look practically identical to me, I think they can be consolidated to one definition in a new method `Context#get_config_path`?

Code in `master` is like this:

```mermaid
classDiagram
`configure.rs#print_configuration` --> `config.rs#StarshipConfig` : in initialize()
`config.rs#StarshipConfig` --> env : config.rs#get_config_path fetches STARSHIP_CONFIG
`context.rs#Context` *-- `config.rs#StarshipConfig` : in new()
```

I think these ideas are ok if the idea is to have all env mocking in `Context`, what do you think?

File: src/context.rs [hunk 2]
```diff
@@ -1,10 +1,11 @@
 use crate::config::{ModuleConfig, StarshipConfig};
 use crate::configs::StarshipRootConfig;
+use crate::context_env::Env;
 use crate::module::Module;
 use crate::utils::{create_command, exec_timeout, read_file, CommandOutput};
 
 use crate::modules;
-use crate::utils::{self, home_dir};
+use crate::utils::{self};
```
Comment (davidkna) [inline src/context.rs:8]:
```suggestion
use crate::utils;
```

**Review by davidkna — COMMENTED** [review]
I have some suggestions, but this PR looks good for the most part, even if this ended up becoming quite a large change set.

File: src/configure.rs [hunk 3]
```diff
@@ -217,41 +216,26 @@ fn handle_toggle_configuration(doc: &mut Document, name: &str, key: &str) -> Res
...
         .parse::<Document>()
         .expect("Failed to load starship config")
 }
 
-pub fn write_configuration(doc: &Document) {
-    let config_path = get_config_path();
+pub fn write_configuration(context: &Context, doc: &Document) {
+    let config_path = context
+        .get_config_path_os()
+        .expect("able to get config path");
```
Comment (davidkna) [inline src/configure.rs:238]: This isn't really worse than what we had, but for user-facing errors like this, please prefer using `eprintln!` and `exit`, the output of `expect` can be hard to parse.

File: src/configure.rs [hunk 4]
```diff
@@ -260,10 +244,15 @@ pub fn write_configuration(doc: &Document) {
...
+pub fn edit_configuration(
+    context: &Context,
+    editor_override: Option<&str>,
+) -> Result<(), Box<dyn std::error::Error>> {
     // Argument currently only used for testing, but could be used to specify
     // an editor override on the command line.
-    let config_path = get_config_path();
+    let config_path = context
+        .get_config_path_os()
+        .expect("config path required to edit the config");
```
Comment (davidkna) [inline src/configure.rs:255]: This isn't really worse than what we had, but for user-facing errors like this, please prefer using `eprintln!` and `exit`, the output of `expect` can be hard to parse.

**Review by davidkna — CHANGES_REQUESTED** [review]

File: src/configure.rs (hunk 3 above)
Comment (domsleee) [reply src/configure.rs:238]: sure, can do

**Review by davidkna — APPROVED** [review]
LGTM
